Tools/ardupilotwaf/ap_libraries.py

```diff
@@ -0,0 +1,140 @@
+# Copyright (C) 2016  Intel Corporation. All rights reserved.
+#
+# This file is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This file is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Waf tool for Ardupilot libraries. The function bld.ap_library() creates the
+necessary task generators for creating the objects of a library for a vehicle.
+That includes the common objects, which are shared among vehicles. That
+function is used by bld.ap_stlib() and shouldn't need to be called otherwise.
+
+The environment variable AP_LIBRARIES_OBJECTS_KW is a dictionary of keyword
+arguments to be passed to bld.objects() when during the creation of the task
+generators. You can use it to pass extra arguments to that function (although
+some of them will be rewritten, see the implementation for details).
+
+NOTE: This tool is provisional, it shouldn't be necessary when all libraries
+become independent of vehicle. Thus, this tool should contain the only
+*minimal* functionality required and addition of features should be avoided.
+"""
+from waflib import Utils
+from waflib.Configure import conf
+from waflib.TaskGen import before_method, feature
+
+import ardupilotwaf as ap
+
+UTILITY_SOURCE_EXTS = ['utility/' + glob for glob in ap.SOURCE_EXTS]
+
+AP_LIBRARIES_VEHICLE_DEPENDENT_SRC = [
+    'libraries/AP_AccelCal/AP_AccelCal.cpp',
+    'libraries/AP_BattMonitor/AP_BattMonitor.cpp',
+    'libraries/AP_Compass/AP_Compass.cpp',
+    'libraries/AP_HAL_Linux/Storage.cpp',
+    'libraries/AP_HAL_PX4/HAL_PX4_Class.cpp',
+    'libraries/AP_HAL_PX4/Storage.cpp',
+    'libraries/AP_HAL_QURT/mainapp/mainapp.cpp',
+    'libraries/AP_HAL_SITL/SITL_cmdline.cpp',
+    'libraries/AP_HAL_VRBRAIN/HAL_VRBRAIN_Class.cpp',
+    'libraries/AP_HAL_VRBRAIN/Storage.cpp',
+    'libraries/AP_InertialSensor/AP_InertialSensor.cpp',
+    'libraries/AP_Mission/AP_Mission.cpp',
+    'libraries/AP_NavEKF2/AP_NavEKF2.cpp',
+    'libraries/AP_NavEKF/AP_NavEKF_core.cpp',
+    'libraries/AP_NavEKF/AP_NavEKF.cpp',
+    'libraries/AP_Rally/AP_Rally.cpp',
+    'libraries/AP_Scheduler/AP_Scheduler.cpp',
+]
```

Comment:
Keeping a list of source files that **currently** have an ifdef is way too fragile.  Anytime one adds an ifdef it could be a potential breakage in the build system